### DISCUSSION TITLE
@Scheduled task using Reactive Hibernate Resources

### DISCUSSION THREAD

**Discussion by MWatter** [quarkusio/quarkus#27543] [Q&A] [answered]
I am having difficult determining  how to do this in the "reactive" way. 

I understand I need to be running the task on the Vert.x event loop so it has to be annotated with @NonBlocking or return Uni<Void>, but what does the call to the Hibernate Panache repo or the PanacheEntity item look like? 

```
    @Scheduled(every = "10s")
    Uni<Void> findItems() {
        MyItem.<MyItem>listAll().onItem().invoke( myItems -> {
            System.out.println(myItem.size());
        });
        return Uni.createFrom().voidItem();
    }
```
I'm certain that isn't right, but I've also tried every combination of `.transform()` and `.subscribe()` I can think of. Either I get an error about the Session being closed" or a i get a timeout of one sort or another (or both in some cases). 

I'm sure there's a simple pattern that I'm missing - enlighten me. 

I just want to read some things from the database, do a bunch of work on them, and the write them back. It's a fairly straightforward set of tasks.

[a comment by an automated account is not shown]

**Comment by mkouba** [comment]
I'll let the hibernate guys to answer the "reactive way" question but I believe that something like this should work just fine:
```java
@Scheduled(every = "10s")
    Uni<Void> findItems() {
        return MyItem.<MyItem>listAll().onItem().invoke( myItems -> {
            System.out.println(myItem.size());
        }).chain(ignored ->  Uni.createFrom().voidItem());
    }
```
What version of quarkus do you use?

**Comment by DavideD** [comment]
@mkouba answer is correct, but you can simplify it a bit:
```
@Scheduled(every = "10s")
Uni<Void> findItems() {
        return MyItem.<MyItem>listAll().invoke( myItems -> {
            System.out.println(myItem.size());
        }).replaceWithVoid();
}
```

@MWatter, Your original solution doesn't work because you are not chaining the unis. This means that when findAll returns the void, listAll might not been executed yet. The session gets closed when the method returns and so when the query actually runs you have the session closed error.

**Comment by DavideD** [comment]
I'm closing this issue but feel free to ask more questions or reopen it if something is not clear.
Cheers

**Comment by DavideD** [comment]
I would rewirte it as something like this:
```java
    @Scheduled(every = "10s")
    Uni<Void> findItems() {
        return MyItem.<MyItem>listAll()
            .invoke( this::printSize )
            .replaceWithVoid();
    }

    private void printSize(List<MyItem> list) {
        System.out.println( list.size() );
    }
```

But it is a question of taste.

**Comment by MWatter** [comment]
Thank you for the suggestions. I'll give them a shot this morning.

**Comment by MWatter** [comment]
Unfortunately, as I've seen in a few other situations, the above suggestions all result in `java.lang.IllegalStateException: Session/EntityManager is closed`. If I annotate the method with `@ReactiveTransactional` which has worked in a few other cases (a transaction shouldn't be needed for a simple read on the DB, but...), I get a different error,
```
java.lang.IllegalStateException: HR000060: Session is closed
```
At this point, this is probably outside the scope of Quarkus and now well within the Hibernate area, so I'll start looking for answers over there... or perhaps just switch to imperative. Thank you for the help.

**Comment by DavideD** [comment]
I doubt this is something related to Hibernate Reactive. If you can push a test project somewhere, we can have a look.

**Comment by DavideD** [comment, marked as answer]
I've adapted one of the quickstarts to test that it actually works. You can find the project here: https://github.com/DavideD/quarkus-quickstarts/tree/scheduler-latest/hibernate-reactive-panache-quickstart

Running `mvn quarkus:dev` will start the project and print the list size on the terminal.
The bean with the scheduler is here: https://github.com/DavideD/quarkus-quickstarts/blob/scheduler-latest/hibernate-reactive-panache-quickstart/src/main/java/org/acme/hibernate/orm/panache/scheduler/CounterBean.java

**Reply by MWatter** [reply]
That's an extremely generous gift of your time. Thank you. I'll run with that and figure things out.

**Reply by MWatter** [reply]
The is for anyone who happens to run across this in the near future. After @DavideD kindly provided the quickstart (something I should have tried myself, frankly), I made a tiny change to the `CounterBean` 's import to get it to compile and boom; it worked great. 

I then started scouring my project against the quickstart to find the differences. Nothing. My project was in it's infancy so it was just barely more than a quickstart itself and I could literally find zero differences.

Finally, in the pom.xml I saw that the quickstart was using `2.11.3.Final` (release about a week ago) and that I was using `2.11.2.Final`. I assumed that there's no way that my using a stable release that was just one minor version behind could be the problem. Well, that was it. Just to be sure, I went back to the quickstart and set it to 2.11.2.Final and sure enough, it failed with the same error I had been experiencing. :)

I really appreciate the help and guidance folks. I'm looking forward to diving in deeper and making Quarkus part of my toolbelt.

**Reply by mkouba** [reply]
FTR it was probably [this issue](https://github.com/quarkusio/quarkus/pull/27108) that was fixed in 2.11.3.
